web3/decoding.nim

```diff
@@ -0,0 +1,437 @@
+  let value = decoder.decode(uint64)
+  if value in T.low.uint64..T.high.uint64:
+    T(value)
+  else:
+    raise newException(SerializationError, "invalid enum value")
+
+proc decode(decoder: var AbiDecoder, T: type Address): T {.raises: [SerializationError].} =
+  var bytes: array[sizeof(T), byte]
+  let padding = abiSlotSize - sizeof(T)
+
+  try:
+    bytes[0..<sizeof(T)] =(decoder.read(sizeof(T))).toOpenArray(padding, abiSlotSize-1)
+  except IOError as e:
+    raise newException(SerializationError, "Failed to read address: " & e.msg)
+
+  T(bytes)
+
+proc decode[I](decoder: var AbiDecoder, T: type array[I, byte]): T {.raises: [SerializationError].} =
+  var arr: T
+
+  let buf = decoder.read(arr.len)
+  arr[0..<arr.len] = buf[0..<arr.len]
+  checkRightPadding(buf, arr.len, buf.len)
+
+  return arr
+
+proc decode(decoder: var AbiDecoder, T: type seq[byte]): T {.raises: [SerializationError].} =
+  let len = decoder.decode(uint64)
+
+  if len == 0:
+    return T.default
+
+  let buf = decoder.read(len.int)
+  checkRightPadding(buf, len.int, buf.len)
+
+  return buf[0 ..< len]
+
+proc decode(decoder: var AbiDecoder, T: type string): T {.raises: [SerializationError].} =
+  string.fromBytes(decoder.decode(seq[byte]))
+
+proc decode[T: tuple](decoder: var AbiDecoder, _: typedesc[T]): T {.raises: [SerializationError].}
+
+proc decode[T: distinct](decoder: var AbiDecoder, _: type T): T {.raises: [SerializationError].} =
+  T(decoder.decode(distinctBase T))
+
+
+proc decodeCollection[T](decoder: var AbiDecoder, size: Opt[uint64]): seq[T] {.raises: [SerializationError].} =
+  ## When T is dynamic, ABI layout looks like:
+  ## +----------------------------+
+  ## | size of the dynamic array |  <-- 32 (optional ONLY for dynamic arrays)
+  ## +----------------------------+
+  ## | offset to element 0       |  <-- 32
+  ## +----------------------------+
+  ## | offset to element 1       |  <-- 32 + size of encoded element 0
+  ## +----------------------------+
+  ## | ...                        |
+  ## +----------------------------+
+  ## | encoded element 0         |  <-- item at offset 0
+  ## +----------------------------+
+  ## | encoded element 1         |  <-- item at offset 1
+  ## +----------------------------+
+  ## | ...                        |
+  ## +----------------------------+
+  if isDynamic(T):
+    # The size of the static array is passed as an argument to the decoder.
+    # For the dynamic array, the size is should be None,
+    # and the decoder will read the size from the input stream.
+    var len = if size.isNone: decoder.decode(uint64) else: size.get()
+
+    var offsets = newSeq[uint64](len)
+    for i in 0..<len:
+      offsets[i] = decoder.decode(uint64)
+
+    result = newSeq[T](len)
+    for i in 0..<len:
+      let pos = decoder.input.pos()
+      if offsets[i].int > pos:
+        decoder.input.advance(offsets[i].int - pos)
+      result[i] = decoder.decode(T)
+
+    return result
```

Comment:
if `result` is used, we should not use `return` - alternatively, use a separate variable and expression return